File: data/derived_cot/rq1_traces/translation/r1/java_to_py/code_output/UserLoginDB.py
import sqlite3

class UserLoginDB:
    def __init__(self, db_name):
        try:
            self.connection = sqlite3.connect(db_name)
            self.create_table()
        except sqlite3.Error as e:
            print(e)

    def create_table(self):
        create_table_query = "CREATE TABLE IF NOT EXISTS users (username TEXT, password TEXT)"
        try:
            cursor = self.connection.cursor()
            cursor.execute(create_table_query)
            self.connection.commit()
        except sqlite3.Error as e:
            print(e)
        finally:
            cursor.close()

    def insert_user(self, username, password):
        insert_query = "INSERT INTO users (username, password) VALUES (?, ?)"
        try:
            cursor = self.connection.cursor()
            cursor.execute(insert_query, (username, password))
            self.connection.commit()
        except sqlite3.Error as e:
            print(e)
        finally:
            cursor.close()

    def search_user_by_username(self, username):
        search_query = "SELECT * FROM users WHERE username = ?"
        try:
            cursor = self.connection.cursor()
            cursor.execute(search_query, (username,))
            row = cursor.fetchone()
            if row:
                return f"{row[0]},{row[1]}"
        except sqlite3.Error as e:
            print(e)
        finally:
            cursor.close()
        return None

    def delete_user_by_username(self, username):
        delete_query = "DELETE FROM users WHERE username = ?"
        try:
            cursor = self.connection.cursor()
            cursor.execute(delete_query, (username,))
            self.connection.commit()
        except sqlite3.Error as e:
            print(e)
        finally:
            cursor.close()

    def validate_user_login(self, username, password):
        user = self.search_user_by_username(username)
        if user:
            parts = user.split(',')
            if len(parts) == 2:
                return parts[1] == password
        return False

    def close(self):
        try:
            if self.connection:
                self.connection.close()
        except sqlite3.Error as e:
            print(e)

File: data/derived_cot/rq1_traces/translation/r1/java_to_cpp/code_output/UserLoginDB.cpp
#include <sqlite3.h>
#include <iostream>
#include <string>
#include <cstring>

class UserLoginDB {
private:
    sqlite3* db;

    void createTable() {
        if (!db) return;

        const char* createTableQuery = "CREATE TABLE IF NOT EXISTS users (username TEXT, password TEXT)";
        char* errMsg = nullptr;
        int rc = sqlite3_exec(db, createTableQuery, nullptr, nullptr, &errMsg);
        if (rc != SQLITE_OK) {
            std::cerr << "SQL error: " << errMsg << std::endl;
            sqlite3_free(errMsg);
        }
    }

public:
    UserLoginDB(const std::string& dbName) : db(nullptr) {
        int rc = sqlite3_open(dbName.c_str(), &db);
        if (rc != SQLITE_OK) {
            std::cerr << "Cannot open database: " << sqlite3_errmsg(db) << std::endl;
            sqlite3_close(db);
            db = nullptr;
            return;
        }
        createTable();
    }

    ~UserLoginDB() {
        close();
    }

    void insertUser(const std::string& username, const std::string& password) {
        if (!db) return;

        const char* insertQuery = "INSERT INTO users (username, password) VALUES (?, ?)";
        sqlite3_stmt* stmt;
        int rc = sqlite3_prepare_v2(db, insertQuery, -1, &stmt, nullptr);
        if (rc != SQLITE_OK) {
            std::cerr << "Failed to prepare statement: " << sqlite3_errmsg(db) << std::endl;
            return;
        }

        rc = sqlite3_bind_text(stmt, 1, username.c_str(), -1, SQLITE_STATIC);
        if (rc != SQLITE_OK) {
            std::cerr << "Failed to bind username: " << sqlite3_errmsg(db) << std::endl;
            sqlite3_finalize(stmt);
            return;
        }

        rc = sqlite3_bind_text(stmt, 2, password.c_str(), -1, SQLITE_STATIC);
        if (rc != SQLITE_OK) {
            std::cerr << "Failed to bind password: " << sqlite3_errmsg(db) << std::endl;
            sqlite3_finalize(stmt);
            return;
        }

        rc = sqlite3_step(stmt);
        if (rc != SQLITE_DONE) {
            std::cerr << "Insert failed: " << sqlite3_errmsg(db) << std::endl;
        }

        sqlite3_finalize(stmt);
    }

    std::string searchUserByUsername(const std::string& username) {
        if (!db) return "";

        const char* searchQuery = "SELECT * FROM users WHERE username = ?";
        sqlite3_stmt* stmt;
        int rc = sqlite3_prepare_v2(db, searchQuery, -1, &stmt, nullptr);
        if (rc != SQLITE_OK) {
            std::cerr << "Failed to prepare statement: " << sqlite3_errmsg(db) << std::endl;
            return "";
        }

        rc = sqlite3_bind_text(stmt, 1, username.c_str(), -1, SQLITE_STATIC);
        if (rc != SQLITE_OK) {
            std::cerr << "Failed to bind username: " << sqlite3_errmsg(db) << std::endl;
            sqlite3_finalize(stmt);
            return "";
        }

        rc = sqlite3_step(stmt);
        if (rc == SQLITE_ROW) {
            const char* u = reinterpret_cast<const char*>(sqlite3_column_text(stmt, 0));
            const char* p = reinterpret_cast<const char*>(sqlite3_column_text(stmt, 1));
            std::string result = std::string(u) + "," + std::string(p);
            sqlite3_finalize(stmt);
            return result;
        } else if (rc == SQLITE_DONE) {
            sqlite3_finalize(stmt);
            return "";
        } else {
            std::cerr << "Search failed: " << sqlite3_errmsg(db) << std::endl;
            sqlite3_finalize(stmt);
            return "";
        }
    }

    void deleteUserByUsername(const std::string& username) {
        if (!db) return;

        const char* deleteQuery = "DELETE FROM users WHERE username = ?";
        sqlite3_stmt* stmt;
        int rc = sqlite3_prepare_v2(db, deleteQuery, -1, &stmt, nullptr);
        if (rc != SQLITE_OK) {
            std::cerr << "Failed to prepare statement: " << sqlite3_errmsg(db) << std::endl;
            return;
        }

        rc = sqlite3_bind_text(stmt, 1, username.c_str(), -1, SQLITE_STATIC);
        if (rc != SQLITE_OK) {
            std::cerr << "Failed to bind username: " << sqlite3_errmsg(db) << std::endl;
            sqlite3_finalize(stmt);
            return;
        }

        rc = sqlite3_step(stmt);
        if (rc != SQLITE_DONE) {
            std::cerr << "Delete failed: " << sqlite3_errmsg(db) << std::endl;
        }

        sqlite3_finalize(stmt);
    }

    bool validateUserLogin(const std::string& username, const std::string& password) {
        std::string user = searchUserByUsername(username);
        if (!user.empty()) {
            size_t pos = user.find(',');
            if (pos == std::string::npos) {
                return false;
            }
            std::string storedPassword = user.substr(pos + 1);
            return storedPassword == password;
        }
        return false;
    }

    void close() {
        if (db) {
            sqlite3_close(db);
            db = nullptr;
        }
    }
};